### PR TITLE
Social: Change publicize endpoint to handle async

### PR DESCRIPTION
**Pull request by gmjuhasz**
Fixes https://github.com/Automattic/jetpack-reach/issues/533

## Proposed changes:

* Made the resharing async

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Testing instructions:

* You need to sandbox this change onto your wpcom sandbox to be able to test this, run `bin/jetpack-downloader test jetpack add/social-publicize-endpoint-async-option` on your Sandbox, and point your JT site to it.
* In your sandbox run the watcher, and watch the logs. Also add `async_reshare_post_sandboxed` to your `0-sandbox` file
* Go to an already published post, choose a connection, then click share post.
* You should see `Your post will be shared on the selected accounts.` in the bottom left
* In your watcher you should see `async_reshare_post_sandboxed` being queued.
* Check that the share is successful on the social media site.

### CODE REVIEW

File: projects/js-packages/publicize-components/src/components/share-post/index.js [hunk 1]
```diff
@@ -33,7 +33,7 @@ function showErrorNotice( message = __( 'Unable to share the Post', 'jetpack' )
  */
 function showSuccessNotice() {
 	const { createSuccessNotice } = dispatch( noticesStore );
-	createSuccessNotice( __( 'Post shared', 'jetpack' ), {
```
Comment (spsiddarthan) [inline projects/js-packages/publicize-components/src/components/share-post/index.js:36]: Nitpick: Your post will be shared in a few seconds.

File: projects/js-packages/publicize-components/src/components/share-post/index.js [hunk 2]
```diff
@@ -33,7 +33,7 @@ function showErrorNotice( message = __( 'Unable to share the Post', 'jetpack' )
...
+	createSuccessNotice( __( 'Your post will be shared on the selected accounts.', 'jetpack' ), {
```
Comment (spsiddarthan) [inline projects/js-packages/publicize-components/src/components/share-post/index.js]:
```suggestion
	createSuccessNotice( __( 'Your post will be shared in a few seconds.', 'jetpack' ), {
```

**Review by spsiddarthan — COMMENTED** [review]
It tests well. 

I was thinking if we should make this better `$async = $request->get_param( 'async' );` but then if we had `$sync = $request->get_param( 'sync' );`, the logic to set default value for sync as true will get complicated. 

We can live with it for now. I have made one comment suggestion, we can make that change and get it merged.

**Comment by spsiddarthan** [comment]
A few additional things to test:
- This branch on Jetpack/Jetpack Social vs Sandbox trunk
- Bleeding edge of Jetpack/Socail vs this branch on the WPCOM sandbox

Seems ok, but I have been jumping across things and want to be sure we test these as well.

File: projects/js-packages/publicize-components/src/components/share-post/index.js (hunk 2 above)
Comment (manzoorwanijk) [reply projects/js-packages/publicize-components/src/components/share-post/index.js]: What about "Your post will be shared soon."?

File: projects/packages/publicize/src/class-rest-controller.php [hunk 3]
```diff
@@ -454,6 +454,9 @@ public function share_post( $request ) {
 		$post_id             = $request->get_param( 'postId' );
 		$message             = trim( $request->get_param( 'message' ) );
 		$skip_connection_ids = $request->get_param( 'skipped_connections' );
+		$async               = $request->get_param( 'async' );
+
+		$is_async_share = isset( $async ) && true === $async;
```
Comment (manzoorwanijk) [inline projects/packages/publicize/src/class-rest-controller.php]: I think we can simply use this

```suggestion
		$async               = (bool) $request->get_param( 'async' );
```

Then update the variable name below.

File: projects/packages/publicize/src/class-rest-controller.php [hunk 4]
```diff
@@ -477,6 +480,7 @@ public function share_post( $request ) {
 			array(
 				'message'             => $message,
 				'skipped_connections' => $skip_connection_ids,
+				'async'               => $is_async_share,
```
Comment (manzoorwanijk) [inline projects/packages/publicize/src/class-rest-controller.php:483]:
```suggestion
				'async'               => $async,
```

File: projects/plugins/jetpack/_inc/lib/core-api/wpcom-endpoints/class-wpcom-rest-api-v2-endpoint-publicize-share-post.php [hunk 5]
```diff
@@ -108,6 +108,9 @@ public function share_post( $request ) {
 		$post_id             = $request->get_param( 'postId' );
 		$message             = trim( $request->get_param( 'message' ) );
 		$skip_connection_ids = $request->get_param( 'skipped_connections' );
+		$async               = $request->get_param( 'async' );
+
+		$is_async_share = isset( $async ) && true === $async;
```
Comment (manzoorwanijk) [inline projects/plugins/jetpack/_inc/lib/core-api/wpcom-endpoints/class-wpcom-rest-api-v2-endpoint-publicize-share-post.php]: Same here

```suggestion
		$async               = (bool) $request->get_param( 'async' );
```

File: projects/plugins/jetpack/_inc/lib/core-api/wpcom-endpoints/class-wpcom-rest-api-v2-endpoint-publicize-share-post.php [hunk 6]
```diff
@@ -120,14 +123,14 @@ public function share_post( $request ) {
 			}
 
 			$publicize = publicize_init();
-			$result    = $publicize->republicize_post( (int) $post_id, $message, $skip_connection_ids, true );
+			$result    = $publicize->republicize_post( (int) $post_id, $message, $skip_connection_ids, true, ! $is_async_share );
```
Comment (manzoorwanijk) [inline projects/plugins/jetpack/_inc/lib/core-api/wpcom-endpoints/class-wpcom-rest-api-v2-endpoint-publicize-share-post.php:126]:
```suggestion
			$result    = $publicize->republicize_post( (int) $post_id, $message, $skip_connection_ids, true, ! $async );
```

**Review by manzoorwanijk — APPROVED** [review]
Works as expected

**Review by spsiddarthan — APPROVED** [review]
Tests well

**Pull request merged**